stan/math/prim/scal/fun/grad_F32.hpp

```diff
@@ -1,57 +1,127 @@
+     * to within <code>precision</code> or throwing when the
+     * function takes <code>max_steps</code> steps.
+     *
+     * This power-series representation converges for all gradients
+     * under the same conditions as the 3F2 function itself.
+     *
+     * @tparam T type of arguments and result
+     * @param[out] g g pointer to array of six values of type T, result.
+     * @param[in] a1 a1 see generalized hypergeometric function definition.
+     * @param[in] a2 a2 see generalized hypergeometric function definition.
+     * @param[in] a3 a3 see generalized hypergeometric function definition.
+     * @param[in] b1 b1 see generalized hypergeometric function definition.
+     * @param[in] b2 b2 see generalized hypergeometric function definition.
+     * @param[in] z z see generalized hypergeometric function definition.
+     * @param[in] precision precision of the infinite sum. defaults to 1e-6
+     * @param[in] max_steps number of steps to take. defaults to 10,000
+     */
     template<typename T>
-    void grad_F32(T* g, T a, T b, T c, T d, T e, T z, T precision = 1e-6) {
+    void grad_F32(T* g, const T& a1, const T& a2, const T& a3, const T& b1,
+        const T& b2, const T& z, const T& precision = 1e-6,
+        int max_steps = 1e5) {
+      check_3F2_converges("grad_F32", a1, a2, a3, b1, b2, z);
+
       using std::log;
       using std::fabs;
       using std::exp;
 
-      T gOld[6];
-
-      for (T *p = g; p != g + 6; ++p) *p = 0;
-      for (T *p = gOld; p != gOld + 6; ++p) *p = 0;
-
-      T tOld = 1;
-      T tNew = 0;
-
-      T logT = 0;
-
-      T logZ = log(z);
-
-      int k = 0;
-
-      while (fabs(tNew) > precision || k == 0) {
-        T C = (a + k) / (d + k);
-        C *= (b + k) / (e + k);
-        C *= (c + k) / (1 + k);
-
-        // If a, b, or c is a negative integer then the series terminates
-        // after a finite number of interations
-        if (C == 0) break;
+      for (T *q = g; q != g + 6; ++q) *q = 0.0;
 
-        logT += (C > 0 ? 1 : -1) * log(fabs(C)) + logZ;
+      T log_g_old[6];
+      for (T *q = log_g_old; q != log_g_old + 6; ++q)
+        *q = -1.0 * std::numeric_limits<double>::infinity();
 
-        tNew = exp(logT);
+      T log_t_old = 0.0;
+      T log_t_new = 0.0;
 
-        gOld[0] = tNew * (gOld[0] / tOld + 1.0 / (a + k));
-        gOld[1] = tNew * (gOld[1] / tOld + 1.0 / (b + k));
-        gOld[2] = tNew * (gOld[2] / tOld + 1.0 / (c + k));
+      T log_z = log(z);
 
-        gOld[3] = tNew * (gOld[3] / tOld - 1.0 / (d + k));
-        gOld[4] = tNew * (gOld[4] / tOld - 1.0 / (e + k));
+      T p = 0.0;
 
-        gOld[5] = tNew * (gOld[5] / tOld + 1.0 / z);
-
-        for (int i = 0; i < 6; ++i) g[i] += gOld[i];
-
-        tOld = tNew;
+      double log_t_new_sign = 1.0;
+      double log_t_old_sign = 1.0;
+      double log_g_old_sign[6];
+      for (double *q = log_g_old_sign; q != log_g_old_sign + 6; ++q)
+        *q = 1.0;
 
+      int k = 0;
+      T term;
```

Comment:
this declaration should be moved down and used in a compound declare/define for `term` where it's defined below.